solidity/contracts/Rewards.sol

```diff
@@ -540,6 +551,55 @@ contract Rewards is Ownable {
         }
     }
 
+    /// @notice Initiates the process of upgrading to another rewards contract.
+    /// @param _newRewardsContract The address of a new rewards contract.
+    function initiateRewardsUpgrade(address _newRewardsContract) public onlyOwner {
+        upgradeInitiatedTimestamp = block.timestamp;
+        newRewardsContract = _newRewardsContract;
+        emit UpgradeInitiated(newRewardsContract);
+    }
+
+    /// @notice Finalizes the process of upgrading to another rewards contract
+    /// by allocating all past intervals and then, transferring the
+    /// not-yet-allocated tokens to a new rewards contract.
+    /// Can be called only when the interval during which the upgrade was
+    /// initiated ended.
+    /// Before finalizing the upgrade, make sure all terminated groups are
+    /// reported.
+    function finalizeRewardsUpgrade() public onlyOwner {
+        require(upgradeInitiatedTimestamp != 0, "Upgrade not initiated");
+
+        // allocate all intervals that can be currently allocated
+        uint256 currentInterval = intervalOf(block.timestamp);
+        uint256 intervalToAllocate = 0;
+        if (currentInterval > 0) {
+            intervalToAllocate = currentInterval.sub(1);
+        }
+        
+        allocateRewards(intervalToAllocate);
+
+        // transfer the unallocated KEEP to the new rewards contract and update
+        // this contract's balances
+        uint256 amountToTransfer = unallocatedRewards;
+
+        totalRewards = totalRewards.sub(amountToTransfer);
+        unallocatedRewards = 0;
+        paidOutRewards = paidOutRewards.add(amountToTransfer);
+
+        emit UpgradeFinalized(amountToTransfer);
+
+        bool success = token.approveAndCall(
+            newRewardsContract,
+            amountToTransfer,
+            bytes("")
+        );
+        if (!success) {
```

Comment:
maybe `require(success, "upgrade finalization failed")`?

Reply:
Sure! More concise is better.